feat(curriculum): add user configuration manager lab

zairahira opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or Gitpod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes https://github.com/freeCodeCamp/CurriculumExpansion/issues/815

<!-- Feel free to add any additional description of changes below this line -->

zairahira committed:
chore: add project structure

zairahira committed:
add hints and tests

zairahira committed:
enhance tests

zairahira committed:
make test flexible

ilenia-magoni commented on client/i18n/locales/english/intro.json:
```diff
@@ -4248,6 +4248,12 @@
         "title": "Certified Full Stack Developer Exam",
         "intro": ["Pass this exam to become a Certified Full Stack Developer."]
+      },
+      "lab-user-configuration-manager": {
+        "title": "Build a User Configuration Manager",
+        "intro": [
+          "In this lab, you will practice working with dictionaries and sets in Python."
+        ]
       }
```
Please replace the entry for `lab-placeholder-dictionaries-and-sets` instead of adding a new property at the end

zairahira committed:
replace placeholder

jdwilkin4 commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,384 @@
+
+1. `add_setting(settings, (key, value))` should:
+
+   - Convert `key` to lowercase.
```
This might be confuse some campers. so far the user stories tell them that the second parameter is a tuple representing a key-value pair. I am not sure how many campers will know to unpack these values from the tuple and then set the key to lowercase.

has there been tons of practice with unpacking values so far in the python curriculum? if not, some might struggle with what `key` is in reference to. 

so maybe more needs to be added here so they know `key` was unpacked from that tuple?

zairahira replied:
Looking at the content prior to this lab, I see there are lecture videos for tuples. The workshop just before it also uses key, value and tuples. I think this should be good to go, but we can still ask @fhsinchy to comment on this.

jdwilkin4 commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,384 @@
+
+**User Stories:**
+
+1. You should create a dictionary named `settings` that would store the following user configuration preferences in this format:
```
I feel like it needs to be clear that this dictionary is only here for testing purposes. 
Because looking at the final solution, this is added at the end only to test out the function.
maybe renaming it would be good. so something like `test_settings`

jdwilkin4 commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,384 @@
+
+# --hints--
+
+You should create a dictionary named `settings` with the correct values.
```
this doesn't check for correct values.
I am able to pass with this code

```py
settings = {
    "not": "anything",
    "it": "something",
    "wassup": "else"
}
```

jdwilkin4 commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,384 @@
+assert.match(code, /settings\s*=\s*{[^}]*}/)
+```
+
+You should define a function named `add_setting`.
````
This passes by only creating the `settings` dictionary.

<img width="953" alt="Screenshot 2025-06-16 at 9 37 23 AM" src="https://github.com/user-attachments/assets/7227ca61-da04-4913-9a86-6c7438d86541" />

jdwilkin4 reviewed (commented):
Went through the lab for the first time and left some comments 👍🏾 

But also, I am not sure if we need to specify that you need to replace `key` in all of these types of strings. and not just return the string exactly. 

```md
Setting 'key' already exists! Cannot add a new setting with this name.
```

zairahira committed:
apply suggestions from code review

zairahira commented:
Hi @jdwilkin4,

I have made the following changes:

• **Added description**: Added introductory text explaining the lab's purpose

• **Restructured user stories**: Improved formatting and clarity of function requirements:
  - Added explicit "Unpack the tuple into a key and value" step
  - Changed message format from `'key'` to `[key]` and `'value'` to `[value]` to indicate that [key] is now a placeholder.

• **Fixed numbering:**
  - The stories as indentation issue that caused the numbering to jumble up.

• `test_setting` is now just supposed to be 'not' empty. Users can add any value as it is just there for testing.

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,379 @@
+})
+```
+
+`add_setting` should return the error message `Setting [key] already exists! Cannot add a new setting with this name.` when the key already exists.
````
The expected message seems to be different from the one shown. I'm referring to the quotes around the key. This happens for other messages too.

zairahira replied:
Changed `[key]` to `'[key]'` and `[value]` to `'[value]'` where applicable.

zairahira committed:
Apply suggestions from code review

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

zairahira committed:
add quotes around key and value

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,379 @@
+})
+```
+
+`add_setting` should return the error message `Setting '[key]' already exists! Cannot add a new setting with this name.` when the key already exists.
````
Usually I'm in favor of using general messages in hints, but here it's a bit confusing. User stories say that the function should convert the key to lowercase but it's not clear where to use the lowercase key. So I would change this hint and the other hints similar to this for the other functions (or add additional hints) into something like this:
```suggestion
`add_setting({'theme': 'light'}, ('THEME': 'dark'))` should return the error message `Setting 'theme' already exists! Cannot add a new setting with this name.`.
```
Also, why not converting to lowercase the value as well?

zairahira replied:
I have updated the existing hints to indicate the dictionary and values passed to functions.

zairahira committed:
update hints

zairahira committed:
tuples were left untested

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,379 @@
+
+1. `add_setting` function should:
+    
+   - Unpack the tuple into a key and value.
```
If the unpacking isn't there, the function would fail itself. I have still added 2 more tests,  I hope it covers the gap.

zairahira replied:
I've removed the requirement to unpack and the related tests.

zairahira committed:
add objective

Dario-DC reviewed (commented):
I left a few comments. Other than those, since this is going to be a cert project, we should add more tests to avoid hard coding.

And I'm not sure if we need to change the challenge type to `23`. I guess that will be discussed later on.

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,406 @@
+   - If the key setting exists, update its value in the `settings` dictionary and return: `Setting '[key]' updated to '[value]' successfully!`
+   - If the key setting doesn't exist, return `Setting '[key]' does not exist! Cannot update a non-existing setting.`
+
+1. You should define a function named `delete_setting` that takes a dictionary of settings and a key as parameters.
```
For definition we should use parameters and not arguments.

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,406 @@
+({ test: () => assert(runPython(`_Node(_code).has_function("add_setting")`)) })
+```
+
+The `add_setting` function should take two parameters.
````
We are talking about the function definition, it should remain parameters.

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,406 @@
+})
+```
+
+The `update_setting` function should take two parameters.
````
We are talking about the function definition, it should remain parameters.

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,406 @@
+})
+```
+
+The `delete_setting` function should take two parameters.
````
We are talking about the function definition, it should remain parameters.

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
````diff
@@ -0,0 +1,406 @@
+})
+```
+
+The `view_settings` function should take one parameter.
````
We are talking about the function definition, it should remain parameters.

zairahira committed:
Apply suggestions from code review

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

zairahira committed:
remove unpacking of variables

zairahira committed:
convert value to lower case and add tests

zairahira committed:
typo

zairahira commented:
Note: I am adding more tests to prevent hard-coding. Do not remove the awaiting update label yet.

zairahira committed:
add additional tests

zairahira committed:
improve wording for parameters

zairahira commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,406 @@
+
+**User Stories:**
+
+1. You should define a function named `add_setting` that takes a dictionary of settings and a tuple containing a key-value pair as parameters.
```
I have applied this style where parameters were asked in the function definition.

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-user-configuration-manager/684aaf9ec670c68d20efd0d0.md:
```diff
@@ -0,0 +1,456 @@
+    test: () => runPython(`
+        settings_in_test = {"theme": "light", "language": "English", "notifications": "enabled"}
+        result = view_settings(settings_in_test)
+        assert result == "Current User Settings:\\nTheme: light\\nLanguage: English\\nNotifications: enabled\\n"
```
The final newline character doesn't seem to be included in the example output. We should either remove it or say explicitly that is required.

zairahira committed:
Apply suggestions from code review

Co-authored-by: Dario <105294544+Dario-DC@users.noreply.github.com>

zairahira committed:
tell about new line

Dario-DC reviewed (commented):
Once merge conflicts are resolved it should be good to go @zairahira

zairahira committed:
Merge branch 'main' into feat/add-user-configuration-manager-lab
